File: tests/c/search_test.c
#include <assert.h>
#include "../../latest/search/search.c"

void test_linear_search() {
    int arr1[] = {1, 2, 3, 4, 5};
    int size1 = 5;

    assert(linear_search(arr1, size1, 3) == 2);  
    assert(linear_search(arr1, size1, 1) == 0);  
    assert(linear_search(arr1, size1, 5) == 4);  
    assert(linear_search(arr1, size1, 10) == -1); 
    assert(linear_search(arr1, size1, -3) == -1);

    int *arr2 = NULL;
    assert(linear_search(arr2, 0, 5) == -1);

}

void test_binary_search() {
    int arr1[] = {1, 2, 3, 4, 5};
    int size1 = 5;

    assert(binary_search(arr1, size1, 3) == 2); 
    assert(binary_search(arr1, size1, 1) == 0);  
    assert(binary_search(arr1, size1, 5) == 4); 
    assert(binary_search(arr1, size1, 10) == -1); 
    assert(binary_search(arr1, size1, -3) == -1);

    int *arr2 = NULL;
    assert(binary_search(arr2, 0, 5) == -1);

}

void test_two_crystal_balls() {
    int arr1[] = {0, 0, 0, 0, 1, 1, 1};
    assert(two_crystal_balls(arr1, 7) == 4);

    int arr2[] = {0, 0, 0, 0, 0, 0, 0};
    assert(two_crystal_balls(arr2, 7) == -1);

    int arr3[] = {1, 1, 1, 1, 1};
    assert(two_crystal_balls(arr3, 5) == 0);

    int arr4[] = {0, 0, 0, 1, 1, 1};
    assert(two_crystal_balls(arr4, 6) == 3);

    int arr5[] = {0, 1, 1, 1, 1};
    assert(two_crystal_balls(arr5, 5) == 1);

    int arr6[] = {0};
    assert(two_crystal_balls(arr6, 0) == -1);

    int *arr7 = NULL;
    assert(two_crystal_balls(arr7, 0) == -1);

    int arr8[] = {0, 0, 0, 0, 0, 1};
    assert(two_crystal_balls(arr8, 6) == 5);

    int arr9[] = {0, 0, 1};
    assert(two_crystal_balls(arr9, 3) == 2);
}

int main() {
    test_linear_search();
    test_binary_search();
    test_two_crystal_balls();
    return 0;
}

File: tests/py/search_test.py
import unittest
from latest.search.search import linear_search, binary_search, two_crystal_balls

class TestSearchFunctions(unittest.TestCase):

    def test_linear_search(self):
        array = [2, 4, 6, 8, 10]
        
        self.assertEqual(linear_search(array, 6), 2)
        self.assertEqual(linear_search(array, 2), 0)
        self.assertEqual(linear_search(array, 10), 4)
        
        self.assertEqual(linear_search(array, 5), -1)
        self.assertEqual(linear_search(array, 11), -1)
        
        self.assertEqual(linear_search([], 3), -1)
        
        self.assertEqual(linear_search([7], 7), 0)
        self.assertEqual(linear_search([7], 2), -1)

    def test_binary_search(self):
        array = [1, 3, 5, 7, 9]
        
        self.assertEqual(binary_search(array, 5), 2)
        self.assertEqual(binary_search(array, 1), 0)
        self.assertEqual(binary_search(array, 9), 4)
        
        self.assertEqual(binary_search(array, 4), -1)
        self.assertEqual(binary_search(array, 10), -1)
        
        self.assertEqual(binary_search([], 3), -1)
        
        self.assertEqual(binary_search([5], 5), 0)
        self.assertEqual(binary_search([5], 1), -1)
    
    def test_two_crystal_balls(self):
        assert two_crystal_balls([False, False, False, True, True]) == 3
        assert two_crystal_balls([False] * 50 + [True] * 50) == 50
        assert two_crystal_balls([True] * 100) == 0
        assert two_crystal_balls([False] * 100) == -1

if __name__ == "__main__":
    unittest.main()
